### ISSUE TITLE
LTX-2 audio decode on Galaxy (4x8/4x32): scaling + tuning — T-pad, BWE channel-TP, blocking re-tune

### ISSUE THREAD

**Issue by smartonTT** [tenstorrent/tt-metal#46423]
## Summary

The LTX-2 audio **vocoder** (`models/tt_dit/models/audio_vae/vocoder_ltx.py`) T-shards its sequence axis across the larger mesh axis. On Galaxy meshes (`4x8`, `4x32`) the T-shard **factor** becomes 8 or 32, and the input-T padding scales as `TILE_HEIGHT * factor`. With the vocoder's tiny input (`t_frames = 120`), this inflates the entire vocoder compute:

| Mesh | T-shard factor | `align = 32*factor` | T padded | Inflation |
|---|---|---|---|---|
| 2x4 | 4 | 128 | 120 → 128 | 1.07x |
| 4x8 | 8 | 256 | 120 → 256 | 2.1x |
| 4x32 | 32 | 1024 | 120 → 1024 | **8.5x** |

Every conv/activation/halo in the vocoder then runs on the padded T. The padding rows propagate at the upsampled length and are cropped from the final waveform, so they are pure waste. On 4x32 the vocoder does ~8.5x more work than necessary.

This was found while optimizing the vocoder on a **bh 2x4 loudbox (8 chips)** in PR #46395. The 2x4 path is fully validated there (4.2x speedup, all `test_audio_components_ltx` cases pass). **The Galaxy behavior below is unvalidated — I had no 32-chip machine.** This issue is for someone with Galaxy access to fix and measure.

## Root cause

`models/tt_dit/models/audio_vae/vocoder_ltx.py` ~line 351-359:

```python
t_pad = 0
if sharded:
    factor = self.parallel_config.factor      # = T-shard factor
    tile_h = 32
    align = tile_h * factor                   # 32 * factor
    rem = x_BTC_torch.shape[1] % align
    if rem != 0:
        t_pad = align - rem
        x_BTC_torch = torch.nn.functional.pad(x_BTC_torch, (0, 0, 0, t_pad))
```

The `32 * factor` alignment exists because the partition is done in **TILE layout** (lines ~363-366):

```python
if sharded:
    x_dev = ttnn.to_layout(x_dev, ttnn.TILE_LAYOUT)
    x_dev = _partition_t(x_dev, self.parallel_config)
    x_dev = ttnn.to_layout(x_dev, ttnn.ROW_MAJOR_LAYOUT)
```

A TILE-layout partition requires each per-chip shard to be tile-aligned (≥32 rows, multiple of 32). With `factor` shards, that forces global T to a multiple of `32*factor`.

The T-shard axis is chosen as the **larger** mesh axis, in two places that must stay consistent:
- Test helper: `models/tt_dit/tests/models/ltx/test_audio_components_ltx.py` ~line 261: `t_axis = 0 if mesh_shape[0] >= mesh_shape[1] else 1`
- Production: `models/tt_dit/pipelines/ltx/pipeline_ltx.py` ~line 1548 (same expression), gated behind `LTX_AUDIO_CHANNEL_TP=1` (channel-TP is currently opt-in; default is single-axis T-shard on the larger axis).

### Fix A (preferred): partition in ROW_MAJOR so per-chip T only needs to be divisible by `factor`, not `32*factor`

If `_partition_t` runs on a ROW_MAJOR tensor, each shard needs only a whole number of rows (T divisible by `factor`), not tile-alignment. For factor=32: pad 120 → 128 (4 rows/chip) instead of 1024. Waste drops from 8.5x to 1.07x.

**Risk / blast radius:** the entire t_pad boundary-correction machinery assumes the current `32*factor` alignment:
- `_tpad_mask` (`models/tt_dit/layers/audio_ops.py` ~line 289)
- `_set_tpad_tail` (`models/tt_dit/layers/audio_ops.py` ~line 310)
- the `_set_tail` closure in `vocoder_ltx.py` forward

These build a validity mask over the global-T tail to zero/replicate the tile-align pad image so the sharded path bit-matches the unsharded path (this guards against an audible "click" at the shard boundary — see the docstring on `_assert_sharded_matches_unsharded` in the test). Any change here MUST keep `test_stage_b_vocoder`'s `sharded vs unsharded max|Δ| < 5e-3` gate green, and ideally keep it at the ~2e-4 level it's at now.

**Must verify:** that `ttnn.mesh_partition` and the downstream sharded ops (`_t_neighbor_pad` halos, `Conv1dViaConv3d`, `depthwise_tap_filter` / `ttnn.experimental.conv1d_depthwise`, the final `_all_gather_t`) all tolerate a non-tile-aligned per-chip T in ROW_MAJOR. Some CCL ops require TILE — check each. If the final all-gather needs TILE, tile-align only there, not for the whole forward.

### Fix B: axis-assignment policy — T-shard the SMALLER axis, channel-TP the LARGER

For 4x32, T-shard factor=4 (align 128, pad 120→128) and channel-TP factor=32 on the size-32 axis. Channel-TP is already implemented and validated at factor=2 (PR #46395, commit `fb7e685487d`): `AudioTCParallelConfig`, `_pick_c_out_block_shard`, `channel_align_unit`.

**Caveat:** channel-TP factor=32 sets `channel_align_unit = 32*32 = 1024`, which pads the late low-channel vocoder stages (96, 48, 24 channels) up to 1024 — a ~10x *channel* padding waste on those stages. The early stages (1536, 768) dominate compute and shard cleanly, so this may still net positive, but it must be measured. The two waste modes (T-pad vs channel-pad) trade off; only a real 4x32 run settles which axis assignment wins.

If you take Fix B, change BOTH the test helper and `pipeline_ltx.py` so they stay consistent, and verify channel-TP factor=4 (the 4x8 case) first since it's the cheaper validation.

### Fix A+B together is likely best
ROW_MAJOR partition kills the T-pad waste; modest channel-TP (factor 4 on the small axis for 4x32) uses the rest of the mesh without the 1024 channel-pad blowup. Measure combinations.

## How to validate (Galaxy required)

1. Build tt-metal on the Galaxy box (follow `knowledge/recipes/tt-metal/build.md` via tt-buddy, or `bash build_metal.sh`).
2. Branch from PR #46395's head: `smarton/ltx-av-conv-perf` (or main if #46395 has merged). Note: that branch carries a cherry-pick of `kevinmi/conv1d-depthwise` (commit `e813c0a89b6`) — the `ttnn.experimental.conv1d_depthwise` op. If it's already on main, rebase to drop it.
3. **Correctness gate** — run the full audio matrix including Galaxy params:
   ```bash
   LTX_AUDIO_FULL_MATRIX=1 pytest models/tt_dit/tests/models/ltx/test_audio_components_ltx.py \
     -k "4x8 or 4x32" -v --timeout 1800
   ```
   All `test_stage_a/b/c` must pass; watch the `sharded vs unsharded device max|Δ|` log line stays < 5e-3 (currently ~2e-4 on 2x4).
4. **Regression gate** — confirm 2x4 is unchanged:
   ```bash
   pytest models/tt_dit/tests/models/ltx/test_audio_components_ltx.py -v --timeout 900
   ```
   (default runs the two 2x4 params; 6 pass.)
5. **Perf measurement** — there is a category profiler harness in the branch at `models/tt_dit/tests/models/ltx/prof_vocoder_forward.py` (`test_prof_vocoder_categories`). It is currently hardcoded to a 2x4 `AudioTCParallelConfig`; adapt its `pc`/mesh to the Galaxy shape and compare wall-time + per-category device-time before/after the fix. Use `tt:profiler` (tracy) for clean device-FW numbers if the harness's host-timed numbers are noisy.

## Acceptance criteria
- `test_audio_components_ltx` passes on `4x8` and `4x32` with `max|Δ| < 5e-3`.
- 2x4 results bit-unchanged (same pass set, `max|Δ|` ~2e-4).
- Vocoder forward wall-time on 4x32 improves materially vs the current `32*factor` t_pad path (expect multiple-x, given the 8.5x inflation).

## Context / provenance
- PR with the validated 2x4 work: #46395 (branch `smarton/ltx-av-conv-perf`).
- The vocoder optimizations already landed (depthwise op, conv blockings, channel-TP enablement, polyphase UpSample, fused SnakeBeta) took the 2x4 vocoder forward from ~1489 ms → ~354 ms.
- Channel-TP machinery to reuse: commit `fb7e685487d` ("ltx audio: enable channel-TP on second mesh axis when both axes > 1").

---

## Additional Galaxy opportunities (from PR #46395 — measured on bh 2x4 loudbox)

The full audio decode (`decode_audio` = mel-VAE + main vocoder + BWE), warm, on the real ~6s clip (601 mel frames), bh 2x4: **MAC 2797 ms → conv1d 1250 ms (2.24x)**. Of the 1250 ms, the **BWE generator alone is ~680 ms** and is compute-bound (see below). Galaxy is the lever for the audio compute, not trace.

### 1. BWE generator channel-TP — `LTX_BWE_CHANNEL_TP=1` (the big audio-decode lever)
The BWE generator is a second full vocoder (24k→48k). `pipeline_ltx.py` `_new_audio_decoder` now **auto-channel-TPs it when the channel factor > 2** (i.e. on 4x8/4x32 but not 2x4); `LTX_BWE_CHANNEL_TP=0/1` overrides. Findings on 2x4:
- Channel-TP on the BWE generator is **bit-identical to single-axis on the real checkpoint weights** (0.0 per-1s-interval vs single-axis). The previously-reported "full-pipeline divergence" is **stale** — fixed by the main-vocoder channel-TP work.
- But it is **+135 ms on 2x4** (1369 vs 1234 ms): the per-conv all-gather tax exceeds the channel-compute split on a 2-wide channel axis.
- **On Galaxy** (channel factor 4) it auto-enables; the split *should* outweigh the gather but **factor 4 is unmeasured** — confirm it's actually net-positive (compare `LTX_BWE_CHANNEL_TP=0` vs `=1` on 4x8/4x32). If factor 4 still loses, raise the `factor > 2` cut. Same `channel_align_unit = 32*factor` low-channel-pad caveat as Fix B applies to the BWE generator's late stages (256→…→16ch).

### 2. Main + BWE channel-TP factor scaling
Both vocoders' channel-TP factor is 2 on 2x4. On Galaxy a larger channel factor splits more of the (compute-bound) conv work, but trades against the `32*factor` channel-pad on low-channel stages. Sweep factor (4, 8) and the T-vs-C axis assignment (ties into Fix B above) — only a real 4x32 run settles it.

### 3. conv3d blocking re-tune for Galaxy shard shapes
`_FP32_BLOCKINGS` (`utils/conv3d.py`) is hand-tuned for the **2x4 per-chip shapes**. On Galaxy the per-chip T/C shard sizes differ, so entries may fall back to the conservative `(32,32,1)` default (`T_out_block=1`, one dispatch/row). The BWE per-conv profiler `test_prof_bwe_per_conv` (in `prof_vocoder_forward.py`) prints each conv's `T_out_block` and whether its shape is tuned — re-run it on the Galaxy shard shapes and add missing entries.

### 4. Not a lever: trace
Trace mode for the vocoder was explored and **closed (#46441)**: at the real 6s length the audio is compute-bound (trace replay = 1251 ms vs eager 1250 ms, bit-identical). Don't spend Galaxy effort on audio trace — spend it on channel-TP (compute split).

### Validation tooling now in the branch
- `test_audio_decode_girl` (checkpoint-gated): runs the real audio decode and reports per-1s-interval error-to-signal (rmse/σ + dB) and PCC **vs a torch oracle** (`_build_torch_stage_c_real`, diffusers vocoder+BWE on CPU with the real checkpoint weights). Use it on Galaxy to confirm channel-TP stays faithful per-second.
- `test_stage_c_audio_decode` (default-on): full VocoderWithBWE at 6s, sharded-vs-unsharded per-window gate.
